hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/common/IncrementalChunkBuffer.java

```diff
@@ -0,0 +1,217 @@
+  private ByteBuffer getAndAllocateAtPosition(int position) {
+    final int i = position / increment;
+    final ByteBuffer ith = getAndAllocateAtIndex(i);
+    Preconditions.checkState(ith.position() == position % increment);
+    return ith;
+  }
+
+  @Override
+  public int position() {
+   // The buffers list must be in the following orders:
+   // full buffers, buffer containing the position, empty buffers, null buffers
+    int i = 0;
+    ByteBuffer ith = null;
+    // buffers must be full
+    for (; i < buffers.size(); i++) {
+      ith = getAtIndex(i);
+      if (ith.position() != increment) {
+        break; // found the first non-full buffer
+      }
+    }
+    // buffer containing the position
+    final int position = i * increment + Optional.ofNullable(getAtIndex(i))
+        .map(ByteBuffer::position).orElse(0);
+    assertRemainingList(ith, i);
+    return position;
+  }
+
+  private void assertRemainingList(ByteBuffer ith, int i) {
+    if (ith != null) {
+      // buffers must be empty
+      for (i++; i < buffers.size(); i++) {
+        ith = getAtIndex(i);
+        if (ith == null) {
+          break; // found the first non-null
+        }
+        Preconditions.checkState(ith.position() == 0);
+      }
+    }
+    // buffers must be null
+    for (i++; i < buffers.size(); i++) {
+      Preconditions.checkState(getAtIndex(i) == null);
+    }
+  }
+
+  @Override
+  public int remaining() {
+    return limit - position();
+  }
+
+  @Override
+  public void clear() {
+    buffers.forEach(ByteBuffer::clear);
+  }
+
+  @Override
+  public void put(ByteBuffer that) {
+    if (that.remaining() > this.remaining()) {
+      final BufferOverflowException boe = new BufferOverflowException();
+      boe.initCause(new IllegalArgumentException(
+          "Failed to put since that.remaining() = " + that.remaining()
+              + " > this.remaining() = " + this.remaining()));
+      throw boe;
+    }
+
+    final int thatLimit = that.limit();
+    int p = position();
+    for (; that.position() < thatLimit;) {
+      final ByteBuffer b = getAndAllocateAtPosition(p);
+      final int min = Math.min(b.remaining(), thatLimit - that.position());
+      that.limit(that.position() + min);
+      b.put(that);
+      p += min;
+    }
+  }
+
+  @Override
+  public ChunkBuffer duplicate(int newPosition, int newLimit) {
+    Preconditions.checkArgument(newPosition >= 0);
+    Preconditions.checkArgument(newPosition <= newLimit);
+    final IncrementalChunkBuffer duplicated = new IncrementalChunkBuffer(
+        newLimit, increment);
+
+    final int pi = newPosition / increment;
+    final int pr = newPosition % increment;
+    final int li = newLimit / increment;
+    final int lr = newLimit % increment;
+
+    for (int i = 0; i < buffers.size(); i++) {
+      final int pos = i < pi ? increment : i == pi ? pr : 0;
+      final int lim = i < li ? increment : i == li ? lr : 0;
+      duplicated.buffers.add(duplicate(i, pos, lim));
+    }
```

Comment:
Do we need to create duplicate buffers (with `lim = 0`) for `i > li`?

Reply:
The private duplicate(..) method will return null when lim==0.

You inspired me that we should modify the code so that the capacity of the last buffer won't go over limit.  Thanks.

Reply:
> The private duplicate(..) method will return null when lim==0.

I don't think it does that.  It will return `null` only if `i` is beyond its own `size`.  `pos` and `lim` are irrelevant in the `null` case.

But now with the `newSize` calculation this problem is fixed.  Thanks.